activemq-client/src/main/java/org/apache/activemq/util/MarshallingSupport.java

```diff
@@ -310,86 +311,53 @@ public static void writeUTF8(DataOutput dataOut, String text) throws IOException
     }
 
     /**
-     * From: http://svn.apache.org/repos/asf/harmony/enhanced/java/trunk/classlib/modules/luni/src/main/java/java/io/DataOutputStream.java
+     * Inspired by: http://svn.apache.org/repos/asf/harmony/enhanced/java/trunk/classlib/modules/luni/src/main/java/java/io/DataOutputStream.java
      */
-    public static long countUTFBytes(String str) {
+    public static long countUTFBytes(String str) throws UTFDataFormatException {
         int utfCount = 0, length = str.length();
         for (int i = 0; i < length; i++) {
-            int charValue = str.charAt(i);
-            if (charValue > 0 && charValue <= 127) {
+            int codePoint = str.codePointAt(i);
+
+            if (codePoint >= 0 && codePoint <= 127) {
                 utfCount++;
-            } else if (charValue <= 2047) {
+            } else if (codePoint <= 2047) {
                 utfCount += 2;
-            } else {
+            } else if (codePoint <= 65535) {
                 utfCount += 3;
+            } else if (codePoint <= 1114111) {
+                utfCount += 4;
+                i++;
+            } else {
+                throw new UTFDataFormatException();
             }
         }
         return utfCount;
     }
 
-    /**
-     * From: http://svn.apache.org/repos/asf/harmony/enhanced/java/trunk/classlib/modules/luni/src/main/java/java/io/DataOutputStream.java
-     */
-    public static int writeUTFBytesToBuffer(String str, long count,
+    public static int writeUTFBytesToBuffer(String str, int count,
                                      byte[] buffer, int offset) throws IOException {
-        int length = str.length();
-        for (int i = 0; i < length; i++) {
-            int charValue = str.charAt(i);
-            if (charValue > 0 && charValue <= 127) {
-                buffer[offset++] = (byte) charValue;
-            } else if (charValue <= 2047) {
-                buffer[offset++] = (byte) (0xc0 | (0x1f & (charValue >> 6)));
-                buffer[offset++] = (byte) (0x80 | (0x3f & charValue));
-            } else {
-                buffer[offset++] = (byte) (0xe0 | (0x0f & (charValue >> 12)));
-                buffer[offset++] = (byte) (0x80 | (0x3f & (charValue >> 6)));
-                buffer[offset++] = (byte) (0x80 | (0x3f & charValue));
-            }
+        byte[] bytes = str.getBytes(StandardCharsets.UTF_8);
+        if (bytes.length != count) {
```

Comment:
`bytes.length` should be always equal to `count`. instead of removing `count` method parameter, I decided to add this check.